### PR TITLE
Micvbang/cij analyser

### PR DESCRIPTION
**Pull request by micvbang**
PR for discussion on direction

### CODE REVIEW

**Comment by safl** [comment]
This is really shaping up.
A thought, how about adding the unittests as part of the ``cij`` package and make then runnable via module load e.g. ``python  -m cij.selftest", then a cijoe-testcase can easily be added running the Python unittest.

**Comment by micvbang** [comment]
Yeah, that's an excellent idea!

File: modules/cij/__init__.py [hunk 1]
```diff
@@ -52,43 +52,46 @@ def get_time_stamp():
     return ""
 
 
-def info(txt):
+def info(txt, fd=sys.stdout):
```
Comment (safl) [inline modules/cij/__init__.py:55]: Put these changes to the output functions into their own commit describing what they enable

Comment (micvbang) [reply modules/cij/__init__.py:55]: Roger!

**Comment by safl** [comment]
Can you rebase on latest and squash linter-fixes?
Also, for "cij: add output file to log functions", this is great, but nothing is using that behavior, right?
I guess it should be possible to either provide a flag telling cijoe to dump the output log inside the output-directory, or another full-path argument to allow the outmessaging to go into a completely separate location.
Or how about, "--log-file" taking a path, and if it is a filename-only then place it relative to results output.

**Comment by micvbang** [comment]
See inline answers

> Can you rebase on latest and squash linter-fixes?

Yep!

> Also, for "cij: add output file to log functions", this is great, but nothing is using that behavior, right?
Yeah, I'm using it [here](https://github.com/refenv/cijoe/pull/16/files#diff-be0e4a41bcac7c70ae06680b13bfb22aca09d5a42d5c7d3960f859ece497a092R176) to format the log-output generated by the analyser, which the reporter reads and includes in the runlogs [here](https://github.com/refenv/cijoe/pull/16/files#diff-769ccd09ca0b6da17cd20e96d9cd11ad003d362d5900caa6429a4a4b0eb1f4d9R92)

> I guess it should be possible to either provide a flag telling cijoe to dump the output log inside the output-directory, or another full-path argument to allow the outmessaging to go into a completely separate location.
> Or how about, "--log-file" taking a path, and if it is a filename-only then place it relative to results output.

**Comment by micvbang** [comment]
Depending on whether we do or don't like there to be coupling between the analyser and the reporter (as there is in this PR right now) I guess we could instead add another field in the trun, e.g. `trun.testcases[].tcase[].extra_runlogs`, to tell the reporter to include these files. That would remove the need for the reporter to explicitly know about the plog format

**Comment by safl** [comment]
> Depending on whether we do or don't like there to be coupling between the analyser and the reporter (as there is in this PR right now) I guess we could instead add another field in the trun, e.g. `trun.testcases[].tcase[].extra_runlogs`, to tell the reporter to include these files in the runlog for that tcase. That would remove the need for the reporter to explicitly know about plog files and allow us to put other stuff in the runlogs without having to change the reporter again in the future

It would be preferable that the postprocessing stuff simply produces its output-files in the _aux folder

**Comment by micvbang** [comment]
Alright, makes sense. I think it would be nice to coordinate the last steps of this on a call :)

**Comment by safl** [comment]
Let's do that :)

File: modules/cij/__init__.py [hunk 2]
```diff
@@ -81,6 +81,7 @@ def err(txt, file=sys.stdout):
     print("%s# %s%s%s" % (PR_ERR_CC, get_time_stamp(), txt, PR_NC), file=file)
     _flush(file)
 
+
```
Comment (safl) [inline modules/cij/__init__.py:84]: This is probably nit-picking, but these should go into the first commit as that is where the spaces where removed in the first-place :)

File: modules/cij/__init__.py [hunk 3]
```diff
@@ -91,6 +92,7 @@ def emph(txt, rval=None, file=sys.stdout):
     else:               # any other value, considered 'bad'
         err(txt, file=file)
 
+
```
Comment (safl) [inline modules/cij/__init__.py:95]: Same here

File: modules/cij/analyser.py [hunk 4]
```diff
@@ -16,6 +16,34 @@
...
 
+_units = {
+    # general
+    '': 1,          # no unit
+    'B': 1,         # bytes
+    'k': 1000,      # kilo
+    'M': 1000**2,   # mega
+    'G': 1000**3,   # giga
+
+    # kibi
```
Comment (safl) [inline modules/cij/analyser.py:27]: This is awesome. But I think that it should use the IEC prefixes for the 1024^n and the SI for the 1000^n as described here: https://en.wikipedia.org/wiki/Binary_prefix

File: modules/cij/analyser.py [hunk 5]
```diff
@@ -16,6 +16,34 @@
...
+    'KB': 1024**1,  # kibibytes
+    'MB': 1024**2,  # mibibytes
+    'GB': 1024**3,  # gibibytes
+    'TB': 1024**4,  # tibibytes
+
+    # kilo
+    'kB': 1000**1,  # kilobytes
+    'mB': 1000**2,  # megabytes
```
Comment (safl) [inline modules/cij/analyser.py:35]: MB

File: modules/cij/analyser.py [hunk 6]
```diff
@@ -16,6 +16,34 @@
...
+_units = {
+    # general
+    '': 1,          # no unit
+    'B': 1,         # bytes
+    'k': 1000,      # kilo
+    'M': 1000**2,   # mega
+    'G': 1000**3,   # giga
+
+    # kibi
+    'KB': 1024**1,  # kibibytes
```
Comment (safl) [inline modules/cij/analyser.py:28]: KiB, MiB, GiB, TiB

File: modules/cij/analyser.py (hunk 4 above)
Comment (micvbang) [reply modules/cij/analyser.py:27]: Good point - this is fixed now

**Comment by safl** [comment]
For the commit message: ``mod:cij:reporter: extract rehome to util `` please change it to:

``Fix rehome() for dataclasses and move to cij.util``

File: modules/cij/extractor.py [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+        return 1
+
+    try:
+        extractor = importlib.util.module_from_spec(spec)
+        spec.loader.exec_module(extractor)
+    except FileNotFoundError:
+        cij.err(f"Cannot find: {name}")
+        raise InitializationError(f"extractor '{name}' not recognized")
+
+    if not getattr(extractor, 'extract_metrics', None):
```
Comment (safl) [inline modules/cij/extractor.py:62]: Great.

File: modules/cij/extractors/__init__.py [hunk 8]
```diff
@@ -0,0 +1 @@
+from cij.extractors import fio_json_read
```
Comment (safl) [inline modules/cij/extractors/__init__.py:1]: Why is this needed?

File: modules/cij/analyser.py [hunk 9]
```diff
@@ -0,0 +1,297 @@
...
+
+
+def preqs_from_file(fpath):
+    """ Read yaml-formatted performance requirements from fpath """
+
+    with open(fpath, 'r') as preqf:
+        return yaml.safe_load(preqf)
+
+
+def plog_from_file(fpath):
```
Comment (safl) [inline modules/cij/analyser.py:143]: The "plog" is deprecated now, right?

Comment (safl) [reply modules/cij/analyser.py:143]: Or is that the "analysis" log?

File: testrequirements/example_06_prequirements.preq [hunk 10]
```diff
@@ -0,0 +1,6 @@
+global:
```
Comment (safl) [inline testrequirements/example_06_prequirements.preq:1]: Does it support this:

```
global:
  iops: "[0;500]k"
  bwps: "[80;200]MiB"
  lat:  "[20;50]usec" 

testcases:
  xnvme_fioe:
    iops: "[0;800]k"
```

Comment (safl) [reply testrequirements/example_06_prequirements.preq:1]: That is, the global will be the default to compare all testcases to, then when "testcases" is provided then for specific testcases the metrics are overwritten.

**Review by safl — COMMENTED** [review]
Looking great, a couple of questions :)

File: modules/cij/analyser.py (hunk 9 above)
Comment (micvbang) [reply modules/cij/analyser.py:142]: Yep, good catch!

File: modules/cij/extractors/__init__.py (hunk 8 above)
Comment (micvbang) [reply modules/cij/extractors/__init__.py:1]: Hmm, that's a good question. I'm using it to make `fio_json_read` available as an import from `cij.extractors`. It's quite possible that we can remove this, but then I think we'll have to change the way that imports of the extractors work.

It would be quite nice to not have to add an import for each extractor we add - it's very easy to forget

File: testrequirements/example_06_prequirements.preq (hunk 10 above)
Comment (micvbang) [reply testrequirements/example_06_prequirements.preq:1]: Not yet - forgot to look at that after our chat yesterday. But it's definitely a nice feature. It shouldn't be difficult to add

File: modules/cij/extractors/__init__.py (hunk 8 above)
Comment (safl) [reply modules/cij/extractors/__init__.py:1]: Yeah, the not so pretty thing to do would be in ``__init__`` doing:
``from cij.extractors import *``

Some PEP will probably complain about that :)

File: testrequirements/example_06_prequirements.preq (hunk 10 above)
Comment (safl) [reply testrequirements/example_06_prequirements.preq:1]: Would be great!

Comment (micvbang) [reply testrequirements/example_06_prequirements.preq:2]: Fixed

File: modules/cij/extractors/fio_json_read.py [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+                pmetric,
+                extr_name=_MY_NAME,
+                fname=os.path.basename(fpath),
+                job_id=n,
+            )
+
+            metrics.append({
+                "ctx": ctx,
+                "iops": job["read"]["iops_mean"],
+                "bwps": job["read"]["bw_mean"] * 1024,
```
Comment (safl) [inline modules/cij/extractors/fio_json_read.py:39]: Perhaps this should be some fancy function utilizing the units?

Comment (safl) [reply modules/cij/extractors/fio_json_read.py:39]: E.g. calling some metric-function saying "I provide **x** in unit **y** give me **x** in default-unit"

Comment (micvbang) [reply modules/cij/extractors/fio_json_read.py:39]: That would be awesome. Should I give an implementation a try?

I'm thinking something like what we discussed the other day: `to_base_unit(bwps, 'KiB')` which would then convert `bwps` to bytes (since that's out base unit for storage)

Comment (micvbang) [reply modules/cij/extractors/fio_json_read.py:39]: Hehe, you got here before me. Sounds like we're thinking the same thing!

Comment (safl) [reply modules/cij/extractors/fio_json_read.py:39]: haha awesome, ``to_base_unit(bwps, 'KiB')`` sounds great!

File: modules/cij/extractors/__init__.py (hunk 8 above)
Comment (micvbang) [reply modules/cij/extractors/__init__.py]: This should be fixed now - changed the way I'm listing extractors

File: modules/cij/extractors/fio_json_read.py (hunk 11 above)
Comment (micvbang) [reply modules/cij/extractors/fio_json_read.py]: This should be done now

Comment (safl) [reply modules/cij/extractors/fio_json_read.py]: cool

File: modules/cij/extractors/fio_json.py [hunk 12]
```diff
@@ -31,6 +32,7 @@ def make_context(fio_obj: dict, fname: str, job_id: int) -> dict:
         "iodepth": opt["iodepth"],
         "fname": fname,
         "job_id": job_id,
+        "extractor_name": extr_name,
```
Comment (safl) [inline modules/cij/extractors/fio_json.py:35]: great!

File: bin/cij_extractor [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+        'output',
+        help="Path to test output",
+        default=os.getcwd()
+    )
+    prsr.register('action', 'custom_extend', cij.util.ExtendAction)
+    prsr.add_argument(
+        "--extractor",
+        help="Name of, or path to, one or more extractor(s)",
+        action="custom_extend",
+        choices=[e.name for e in cij.extractor.find_extractors()],
```
Comment (safl) [inline bin/cij_extractor:42]: Great!

Comment (micvbang) [reply bin/cij_extractor:42]: Thanks! I randomly stumbled upon it and found that it fit very well. It even lists the existing ones if you don't supply one

**Comment by safl** [comment]
Thanks!
